contextualized/regression/datasets.py

```diff
@@ -1,121 +1,120 @@
 """
-Data generators used for Contextualized regression training.
+Datasets used for Contextualized regression training.
 """
 from abc import abstractmethod
 import torch
-from torch.utils.data import IterableDataset
 
 
-class Dataset:
-    """Superclass for datastreams (iterators) used to train contextualized.regression models"""
+class RegressionDatasetBase:
+    """
+    Superclass for map-based datasets used to train contextualized.regression models
+    """
+
+    def __init__(self, C, X, Y, idx=[], dtype=torch.float):
 
-    def __init__(self, C, X, Y, dtype=torch.float):
         self.C = torch.tensor(C, dtype=dtype)
         self.X = torch.tensor(X, dtype=dtype)
         self.Y = torch.tensor(Y, dtype=dtype)
-        self.n_i = 0
-        self.x_i = 0
-        self.y_i = 0
-        self.n = C.shape[0]
         self.c_dim = C.shape[-1]
         self.x_dim = X.shape[-1]
         self.y_dim = Y.shape[-1]
+        self.n = self.C.shape[0]
         self.dtype = dtype
 
-    def __iter__(self):
+        if idx == []:  # use full dataset
+            self.idx = range(self.n)
```

Comment:
Thanks for pointing this out. Before  "_generate_sample_ids" was abstracted, each dataset had its own variation of the line `self.sample_ids = [self.all_sample_ids[i] for i in self.idx]`. For instance something like `self.sample_ids = [s for s in self.all_samples if s[1] in self.idx]` (in the case where length of total samples is n * y_dim, and y_dim=3). This made it easier at the time to have randomization work, because generating random indeces for test/train/val was happening at the datamodule level, but iterating to create additional samples (eg, y_dim * n) was happening at the dataset level. I forgot about this when doing the abstracting. I realize this can easily be replaced by specifying the dataset type in the datamodule and using if statements to create the indeces of the full range beforehand.

Reply:
I think the multi-task indexing logic should only live in the actual datasets (not the datamodule). That way we only have to think about sample-level indexing outside of the datasets.

I think it might make more sense to trim the input C, X, Y matrices according to the sample `idx` in the baseclass `__init__`, and then create the randomized multi-task sample list using these trimmed matrices and the subclass-specific `__len__`.

Reply:
I think this still needs attention. `idx` still needs to select samples before `self.samples` is generated. Also let's use `None` as a default "unspecified" kwarg since [] isn't adding any important functionality.